Reduce unsafeness in FormListedElement

annevk opened the pull request:
#### 36bad955ba304a767177a02ace11ce9f0ca12e76
<pre>
Reduce unsafeness in FormListedElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=304141">https://bugs.webkit.org/show_bug.cgi?id=304141</a>

Reviewed by Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Tested as neutral on Speedometer.

Canonical link: <a href="https://commits.webkit.org/304663@main">https://commits.webkit.org/304663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b714b097aff07ca9bf1f0e207ee422270f0d191f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136247 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8603 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143958 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/737407b2-0a85-4550-a798-2452fea8a4a4) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138119 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8448 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104176 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/40af09ad-85c7-4b72-a71c-1a2c4adb54da) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122085 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85007 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b9e2358d-55e6-464a-b31e-be4b22388562) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/4066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4551 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115693 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146703 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8287 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/40853 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8303 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6960 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112859 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6334 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118390 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20987 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8335 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36447 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8053 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71894 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/8275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/8127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -48,12 +49,12 @@ class FormAttributeTargetObserver final : private IdTargetObserver {
     WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(FormAttributeTargetObserver);
 public:
-    FormAttributeTargetObserver(const AtomString& id, FormListedElement&);
+    explicit FormAttributeTargetObserver(const AtomString& id, FormListedElement&);
```
Why explicit? There is more than one mandatory parameter so it is unusual.

cdumez commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -123,9 +124,9 @@ void FormListedElement::formOwnerRemovedFromTree(const Node& formRoot)
-    Node* rootNode = &asHTMLElement();
-    auto* currentForm = form();
-    for (auto* ancestor = asHTMLElement().parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+    SUPPRESS_UNCOUNTED_LOCAL Node* rootNode = &asHTMLElement();
```
cc @rniwa . I am not convinced we want to suppress here. Odds are we need to make this code safe.

annevk replied:
(I can also remove this for now and replace it with a FIXME pointing to a bug if we're not ready to make a call.)

rniwa replied:
So this entire code block is non-trivial because `asHTMLElement` is out-of-line and so is `setFormInternal`. We could use the new static analyzer feature to ignore the variable declaration inside `if`'s condition when "then" statement is trivial.

annevk replied:
@rniwa sorry, I'm not sure what changes I need to make based on that comment. I'll schedule A/B testing though.

rniwa commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -296,14 +297,14 @@ const AtomString& FormListedElement::name() const
 
 FormAttributeTargetObserver::FormAttributeTargetObserver(const AtomString& id, FormListedElement& element)
-    : IdTargetObserver(element.asHTMLElement().treeScope().idTargetObserverRegistry(), id)
+    : IdTargetObserver(element.asProtectedHTMLElement()->protectedTreeScope()->idTargetObserverRegistry(), id)
```
This is unfortunate. `idTargetObserverRegistry()` is inlined but `ensureIdTargetObserverRegistry()` is out of line...

rniwa reviewed (approved):
Did you perf test this PR? I'm worried this might be a Speedometer regression.

rniwa commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -123,9 +124,9 @@ void FormListedElement::formOwnerRemovedFromTree(const Node& formRoot)
-    for (auto* ancestor = asHTMLElement().parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+    SUPPRESS_UNCOUNTED_LOCAL Node* rootNode = &asHTMLElement();
+    SUPPRESS_UNCOUNTED_LOCAL auto* currentForm = form();
+    for (SUPPRESS_UNCOUNTED_LOCAL auto* ancestor = asHTMLElement().parentNode(); ancestor; ancestor = ancestor->parentNode()) {
```
We can just use `rootNode->parentNode()` here and that would be more efficient than calling this virtual function for the second time.

rniwa commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -123,9 +124,9 @@ void FormListedElement::formOwnerRemovedFromTree(const Node& formRoot)
-    for (auto* ancestor = asHTMLElement().parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+    SUPPRESS_UNCOUNTED_LOCAL Node* rootNode = &asHTMLElement();
+    SUPPRESS_UNCOUNTED_LOCAL auto* currentForm = form();
+    for (SUPPRESS_UNCOUNTED_LOCAL auto* ancestor = rootNode->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
```
I don't think this suppression is necessary since this entire "for" block is trivial?

annevk replied:
It's because of _ancestor_. See https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/304574%40main%20%286732%29/scan-build-output/StaticAnalyzer/WebCore/StaticAnalyzerReports/report-FormListedElement.cpp-formOwnerRemovedFromTree-5-7a0e90.html#EndPath

We also compare _ancestor_ as a pointer below which is unsafe.

rniwa replied:
That's because `asHTMLElement()` in `asHTMLElement().parentNode()` is non-trivial. After your code change, we're no longer calling `asHTMLElement()` so ancestor shouldn't emit any warnings.

annevk replied:
Unfortunately the bot continues to be unhappy about these variables, even with the new strategy: https://ews-build.s3-us-west-2.amazonaws.com/macOS-Safer-CPP-Checks-EWS/85a97837-71686/scan-build-output/StaticAnalyzerRegressions/WebKitLegacy/index.html

annevk committed:
Reduce unsafeness in FormListedElement
https://bugs.webkit.org/show_bug.cgi?id=304141

Reviewed by Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Tested as neutral on Speedometer.

Canonical link: https://commits.webkit.org/304663@main
